### ISSUE TITLE
 Illegal base64 character 2d error occurs if you submit select uuid()

### ISSUE THREAD

**Issue by wyukawa** [prestosql/presto#853]
Hi,

We upgraded prestosql to 312 but the error occurred.

* version
312

* query
```select uuid()```

* error log
```
presto:default> select uuid();
java.lang.RuntimeException: Error fetching next at http://host:8080/v1/statement/executing/20190530_035337_00358_5j6u4/x18968561a7bc4c0890e448ce2f6ae78a/0 returned an invalid response: JsonResponse{statusCode=200, statusMessage=OK, headers={content-type=[application/json], date=[Thu, 30 May 2019 03:53:37 GMT], vary=[Accept-Encoding, User-Agent], x-content-type-options=[nosniff]}, hasValue=false} [Error: {"id":"20190530_035337_00358_5j6u4","infoUri":"http://host:8080/ui/query.html?20190530_035337_00358_5j6u4","partialCancelUri":"http://host:8080/v1/stage/20190530_035337_00358_5j6u4.0","nextUri":"http://host:8080/v1/statement/executing/20190530_035337_00358_5j6u4/x18968561a7bc4c0890e448ce2f6ae78a/1","columns":[{"name":"_col0","type":"uuid","typeSignature":{"rawType":"uuid","typeArguments":[],"literalArguments":[],"arguments":[]}}],"data":[["8bbf4e4f-1e28-494e-bfc3-b1ccfb8e45d6"]],"stats":{"state":"RUNNING","queued":false,"scheduled":true,"nodes":1,"totalSplits":17,"queuedSplits":17,"runningSplits":0,"completedSplits":0,"cpuTimeMillis":0,"wallTimeMillis":0,"queuedTimeMillis":1,"elapsedTimeMillis":38,"processedRows":0,"processedBytes":0,"peakMemoryBytes":0,"spilledBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":17,"queuedSplits":17,"runningSplits":0,"completedSplits":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]},"progressPercentage":0.0},"warnings":[]}
]
	at io.prestosql.client.StatementClientV1.requestFailedException(StatementClientV1.java:463)
	at io.prestosql.client.StatementClientV1.advance(StatementClientV1.java:402)
	at io.prestosql.cli.StatusPrinter.printInitialStatusUpdates(StatusPrinter.java:131)
	at io.prestosql.cli.Query.renderQueryOutput(Query.java:151)
	at io.prestosql.cli.Query.renderOutput(Query.java:136)
	at io.prestosql.cli.Console.process(Console.java:337)
	at io.prestosql.cli.Console.runConsole(Console.java:264)
	at io.prestosql.cli.Console.run(Console.java:154)
	at io.prestosql.cli.Presto.main(Presto.java:31)
Caused by: java.lang.IllegalArgumentException: Unable to create class io.prestosql.client.QueryResults from JSON response:
[{"id":"20190530_035337_00358_5j6u4","infoUri":"http://host:8080/ui/query.html?20190530_035337_00358_5j6u4","partialCancelUri":"http://host:8080/v1/stage/20190530_035337_00358_5j6u4.0","nextUri":"http://host:8080/v1/statement/executing/20190530_035337_00358_5j6u4/x18968561a7bc4c0890e448ce2f6ae78a/1","columns":[{"name":"_col0","type":"uuid","typeSignature":{"rawType":"uuid","typeArguments":[],"literalArguments":[],"arguments":[]}}],"data":[["8bbf4e4f-1e28-494e-bfc3-b1ccfb8e45d6"]],"stats":{"state":"RUNNING","queued":false,"scheduled":true,"nodes":1,"totalSplits":17,"queuedSplits":17,"runningSplits":0,"completedSplits":0,"cpuTimeMillis":0,"wallTimeMillis":0,"queuedTimeMillis":1,"elapsedTimeMillis":38,"processedRows":0,"processedBytes":0,"peakMemoryBytes":0,"spilledBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":17,"queuedSplits":17,"runningSplits":0,"completedSplits":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]},"progressPercentage":0.0},"warnings":[]}
]
	at io.prestosql.client.JsonResponse.<init>(JsonResponse.java:70)
	at io.prestosql.client.JsonResponse.execute(JsonResponse.java:144)
	at io.prestosql.client.StatementClientV1.advance(StatementClientV1.java:388)
	... 7 more
Caused by: java.lang.IllegalArgumentException: Invalid JSON string for [simple type, class io.prestosql.client.QueryResults]
	at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:137)
	at io.prestosql.client.JsonResponse.<init>(JsonResponse.java:67)
	... 9 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `io.prestosql.client.QueryResults`, problem: Illegal base64 character 2d
 at [Source: (String)"{"id":"20190530_035337_00358_5j6u4","infoUri":"http://host:8080/ui/query.html?20190530_035337_00358_5j6u4","partialCancelUri":"http://host:8080/v1/stage/20190530_035337_00358_5j6u4.0","nextUri":"http://host:8080/v1/statement/executing/20190530_035337_00358_5j6u4/x18968561a7bc4c0890e448ce2f6ae78a/1","columns":[{"name":"_col0","type":"uuid","typeSignature":{"rawType":"uuid","typeArguments":[],"literalArguments":[],"arguments":[]}}],"data":[["8bbf4e4f-"[truncated 600 chars]; line: 1, column: 1099]
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1608)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:484)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:503)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:285)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:229)
	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:195)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:488)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4013)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3042)
	at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:134)
	... 10 more
Caused by: java.lang.IllegalArgumentException: Illegal base64 character 2d
	at java.base/java.util.Base64$Decoder.decode0(Base64.java:743)
	at java.base/java.util.Base64$Decoder.decode(Base64.java:535)
	at java.base/java.util.Base64$Decoder.decode(Base64.java:558)
	at io.prestosql.client.FixJsonDataUtils.fixValue(FixJsonDataUtils.java:179)
	at io.prestosql.client.FixJsonDataUtils.fixData(FixJsonDataUtils.java:73)
	at io.prestosql.client.QueryResults.<init>(QueryResults.java:69)
	at jdk.internal.reflect.GeneratedConstructorAccessor13.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.fasterxml.jackson.databind.introspect.AnnotatedConstructor.call(AnnotatedConstructor.java:124)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:283)
	... 19 more
Query is gone (server restarted?)
```

**Comment by electrum** [comment]
Thanks for reporting this error. We forgot to add a mapping in the client when the type was added.

The root cause is that the client protocol uses JSON, which cannot represent binary natively, so binary is encoded as a base64 string. The problem is that the client doesn't know if the JSON string is supposed to be a textual string or encoded binary data. Long ago, we made the choice to have the client assume any unknown type is binary. In retrospect, that seems to be the wrong choice, as the failure mode is an error, rather than giving the user a base64 string.

The seemingly obvious solution is for the server to tell the client if the representation is binary, either directly or via the encoding process, but that turns out to be difficult and complicated.
